### DISCUSSION TITLE
Custom Kubernetes Operator/Controller Rust

### DISCUSSION THREAD

**Discussion by EliXorZz** [kube-rs/kube#954] [Q&A] [answered]
Hello, I'm implementing my own Kubernetes operator in Rust and I was wondering if it was ok to have 2 Controllers in the same project that both watch a different custom resource.

Otherwise I don't really see how to go about it and what features I have to use to achieve what I want!

**Comment by clux** [comment, marked as answer]
heyhey. Yep, it will work, but with one caveat:

We don't currently a controller-manager equivalent here so there's no efficient sharing of streams across controllers. If any of the two controllers is for the same resource, or watch the same child resource, you end up querying the apiserver twice for that resource. This is because they set up their watches independently.

Using the [`applier`](https://docs.rs/kube/latest/kube/runtime/fn.applier.html) directly (i.e. passing in raw streams) allows for doing this more efficiently, but it is a lot less ergonomic.

Been planning on opening an issue for this. We barely [explored some thoughts on this](https://github.com/kube-rs/kube-rs/pull/824#issuecomment-1133769478) semi-recently.

**Comment by EliXorZz** [comment]
There is a resource where I need to keep it in a valid state all the time so I will use one controller, the other I only need to read it at certain time to retrieve information from it, in this case the  it is better to use the Reflector, isn't it?

**Reply by clux** [reply]
ah, yeah, this sounds like a reflector use case.

Note that **if** the thing you need to read elsewhere is the main resource you are using a controller for (the resource passed to `Controller::new`) then you can actually share a reflector with the controller via [`Controller::store`](https://docs.rs/kube/latest/kube/runtime/struct.Controller.html#method.store).

**Reply by EliXorZz** [reply]
thank you for answering so quickly 👍
